### ISSUE TITLE
auto formatting of `run` shouldn't split its content into one word per line

### ISSUE THREAD

**Issue by Khady** [ocaml/dune#10860]
## Desired Behavior

When using `run` in a rule stanza we can have long command with many arguments which quickly end up being split into one word per line. It would tend to look better if `run` could keep everything on a single line as much as possible. And when it splits things into multiple lines it would put as much content as possible on every line.

## Example

The current

```
(rule
 (target sql_file1.ml)
 (deps
  gen_file2.sql
  gen_file3.sql)
 (mode promote)
 (action
  (with-stdout-to
   %{target}
   (run
    sqlgg
    -static-header
    -gen
    caml_io
    -name
    Make
    -params
    unnamed
    -category
    none
    gen_file2.sql
    -category
    all
    gen_file3.sql))))
```

could look more like this

```lisp
(rule
 (target sql_file1.ml)
 (deps 
  gen_file2.sql
  gen_file3s.sql)
 (mode promote)
 (action
  (with-stdout-to
   %{target}
   (run  sqlgg -static-header -gen caml_io -name Make -params unnamed -category none gen_file2.sql -category all gen_file3.sql))))
```

or if the line is longer than the limit, it could look like this
```
   (run  sqlgg -static-header -gen caml_io -name Make -params unnamed -category none 
            gen_file2.sql -category all gen_file3.sql))))
```

**Comment by rgrinberg** [comment]
> or if the line is longer than the limit, it could look like this

This option seems like a strict improvement over what we're currently doing.

Would you like to send a PR to fix this? Note that the behavior will need to be versioned.

**Comment by nojb** [comment]
See #10892

**Comment by Khady** [comment]
IIUC this one was implemented but then reverted. What are the conditions to make it happen again?

**Comment by Khady** [comment]
And can we re-open this issue so that it doesn't get lost?

**Comment by nojb** [comment]
> IIUC this one was implemented but then reverted. What are the conditions to make it happen again?

Hello, thanks for the ping. I should have explained what happened to this change here.

- The change itself was made in #10892, and made available for Dune language version >= 3.17
- The original change was reverted from the 3.17 release in #11164 because it caused the breakage of a test in `odoc` which used `dune format-dune-file` (which always uses the latest version of the formatter if a specific version is not specified)
- The change was kept in the main development branch and slated to be released with 3.18
- The version guard of the formatting change was fixed from 3.17 to 3.18 in #11175
- As an alternative to the use of `dune format-dune-file` in the `odoc` testsuite, an action `(format-dune-file)` (which obeys the current Dune project language version) was exposed in #11166, made available for Dune language version >= 3.18
- The version guard of the formatting change was bumped from 3.18 to 3.19 in #11486 to give time to users of `dune format-dune-file` in Dune files (eg `odoc`) to switch to `(format-dune-file)` before introducing a breaking change. After 3.18 is released the `odoc` maintainers should be contacted to switch to `(format-dune-file)`

TL;DR: the change has not been reverted, it exists in `main`, but will only be available in 3.19 and later.

> And can we re-open this issue so that it doesn't get lost?

Since the change has not been reverted, I think we don't necessarily have to re-open the issue. Do you agree?

Apologies though for the long waiting time before this change reaches the public. It was really a bad combination of many factors (as you can devise by reading the long-winded trace of changes above) and Dune's empahsis on not breaking existing users as much as possible.

As to when 3.18 and 3.19 may be released, this is a matter for the release manager (@maiste?).

Cheers!

**Comment by Khady** [comment]
Thanks for all the details!

**Comment by maiste** [comment]
Thanks @nojb for the summary! 😄 
Regarding the release cycle, we try to maintain a ~3 months schedule. I have planned to look for the `3.18` release next week. It should be out before the end of March on the pessimistic version and next week on the optimistic one.

**Comment by Khady** [comment]
https://github.com/ocaml/dune/blob/574d25e2fe5fad19ca2e2d522ad4a9657e00f6d6/src/dune_lang/format.ml#L29

looks like it isn't :(

**Comment by nojb** [comment]
> [dune/src/dune_lang/format.ml](https://github.com/ocaml/dune/blob/574d25e2fe5fad19ca2e2d522ad4a9657e00f6d6/src/dune_lang/format.ml#L29)
> 
> Line 29 in [574d25e](/ocaml/dune/commit/574d25e2fe5fad19ca2e2d522ad4a9657e00f6d6)
> 
>  (if version < (3, 20) then Pp.hvbox else Pp.hovbox) 
> looks like it isn't :(

Hello @Khady, thanks for following up on this. For the record, I also learned of this change after the release, I don't know what was behind it. I guess it is a question for the release manager @maiste.

**Comment by rgrinberg** [comment]
I delayed the change until 3.20. The plan was to delay the change until ocamlformat's test suite was fixed, which we thought was going to happen before 3.20. Since ocamlformat's test suite isn't fixed yet, the change has been delayed to 3.20

**Comment by Khady** [comment]
could be introduced again now that format-dune-file respects dune-project?
